### PR TITLE
Build release tars using bazel

### PR DESCRIPTION
**Pull request by ixdy**
**What this PR does / why we need it**: builds equivalents of the various kubernetes release tarballs, solely using bazel.

For example, you can now do
```console
$ make bazel-release
$ hack/e2e.go -v -up -test -down
```

**Special notes for your reviewer**: this is currently dependent on https://github.com/ixdy/bazel/commit/3b29803eb528ff525c7024190ffbf4b08c598cf2, which I have yet to turn into a pull request, since I'm still trying to figure out if this is the best approach.

Basically, the issue comes up with the way we generate the various server docker image tarfiles and load them on nodes:
* we `md5sum` the binary being encapsulated (e.g. kube-proxy) and save that to `$binary.docker_tag` in the server tarball
* we then build the docker image and tag using that md5sum (e.g. `gcr.io/google_containers/kube-proxy:$MD5SUM`)
* we `docker save` this image, which embeds the full tag in the `$binary.tar` file.
* on cluster startup, we `docker load` these tarballs, which are loaded with the tag that we'd created at build time. the nodes then use the `$binary.docker_tag` file to find the right image.

With the current bazel `docker_build` rule, the tag isn't saved in the docker image tar, so the node is unable to find the image after `docker load`ing it.

My changes to the rule save the tag in the docker image tar, though I don't know if there are subtle issues with it. (Maybe we want to only tag when `--stamp` is given?)

Also, the docker images produced by bazel have the timestamp set to the unix epoch, which is not great for debugging. Might be another thing to change with a `--stamp`.

Long story short, we probably need to follow up with bazel folks on the best way to solve this problem.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by ixdy** [comment]
making a few changes/fixes, sorry for the noise.

I think this should be working now.

**Comment by ixdy** [comment]
> I think this should be working now.

I lied. But maybe now.

**Comment by spxtr** [comment]
@k8s-bot node e2e test this
#35983

File: cluster/saltbase/BUILD [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    visibility = ["//visibility:private"],
+)
+
+filegroup(
+    name = "all-srcs",
+    srcs = [":package-srcs"],
+    tags = ["automanaged"],
+)
+
+# TODO(#3579): This is a temporary hack. It gathers up the yaml,
```
Comment (spxtr) [inline cluster/saltbase/BUILD]: How temporary?

File: build/release-tars/BUILD [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        "//build:docker-artifacts",
+    ],
+    mode = "0755",
+    package_dir = "server/bin",
+    visibility = ["//visibility:private"],
+)
+
+genrule(
+    name = "dummy",
+    outs = [".dummy"],
```
Comment (spxtr) [inline build/release-tars/BUILD]: :confused:

**Comment by spxtr** [comment]
Overall this looks really spiffy. Have you reached out to the bazel people about the docker rule updates?

**Comment by spxtr** [comment]
I'm not too happy the build speeds, but we can optimize later. It's still significantly faster than `make quick-release` for both the initial and incremental builds.

File: build/release-tars/BUILD [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    cmd = "touch $@",
+)
+
+# Some of the startup scripts fail if there isn't an addons/ directory in the server tarball.
```
Comment (mikedanese) [inline build/release-tars/BUILD]: TODOD: delete this

**Comment by mikedanese** [comment]
This is so beautiful i cried

File: build/release-tars/BUILD (hunk 3 above)
Comment (ixdy) [inline build/release-tars/BUILD]: yeah, this part makes me sad. I'm not sure how many places assume `server/addons/` exists, but that can hopefully be addressed in a follow-up.

File: cluster/saltbase/BUILD (hunk 1 above)
Comment (ixdy) [inline cluster/saltbase/BUILD]: I copied the comment from [here](https://github.com/kubernetes/kubernetes/blob/master/build/lib/release.sh#L338-L344). I'm still not sure if it's necessary or not.

Comment (spxtr) [inline cluster/saltbase/BUILD]: Ah, "temporary"

**Comment by ixdy** [comment]
@k8s-bot gke e2e test this

**Pull request merged**